### PR TITLE
sof: add all global structures to main firmware context

### PR DESCRIPTION
**Pull request by tlauda**
This set of patches adds all global structures to main firmware context. It's the first step to achieve easier and more understandable multicore support. Since all of this data can be accessed by many different cores we need to avoid cache synchronization problems by having only one point of access, which we can control.

### CODE REVIEW

File: src/include/sof/sof.h [hunk 1]
```diff
@@ -28,6 +39,40 @@ struct sof {
...
+	/* shared notifier data */
+	struct notify_data *notify_data;
+
+	/* platform dai information */
+	struct dai_info *dai_info;
+
+	/* platform DMA information */
+	struct dma_info *dma_info;
+
+	__aligned(PLATFORM_DCACHE_ALIGN) int alignment[0];
```
Comment (jajanusz) [inline src/include/sof/sof.h]: Why is this needed? Do we want to protect some objects that may come in memory after 'sof' object, so they will not get partially de/cached?

Comment (tlauda) [reply src/include/sof/sof.h]: @jajanusz Exactly. To be 100% sure that slaves core won't mess up the the rest of the data for master core we need to align ending of the sof object to cache line. Currently before slave core enablement we writeback the whole bss.

File: src/platform/library/include/platform/lib/memory.h [hunk 2]
```diff
@@ -10,6 +10,8 @@
 #ifndef __PLATFORM_LIB_MEMORY_H__
 #define __PLATFORM_LIB_MEMORY_H__
 
+#define PLATFORM_DCACHE_ALIGN	sizeof(void *)
```
Comment (lgirdwood) [inline src/platform/library/include/platform/lib/memory.h]: Btw, what are we aligning here ? Are we aligning on data cache line size and wouldn't this be architecture or platform specific ?

File: src/include/sof/sof.h (hunk 1 above)
Comment (lgirdwood) [reply src/include/sof/sof.h]: It's probably worth adding some comments next to the main sof structure explaining this.

File: src/platform/baytrail/include/platform/lib/clk.h [hunk 3]
```diff
@@ -15,6 +15,8 @@
 #include <sof/lib/shim.h>
 #include <stdint.h>
 
+struct sof;
```
Comment (lgirdwood) [inline src/platform/baytrail/include/platform/lib/clk.h]: I wonder if we should just declare this in a global "context" header ?

File: src/platform/library/include/platform/lib/memory.h (hunk 2 above)
Comment (tlauda) [reply src/platform/library/include/platform/lib/memory.h]: @lgirdwood Yes, this is platform specific, so that's why it's in the platform headers. Previously it was placed in the arch for host, which was invalid.

File: src/platform/baytrail/include/platform/lib/clk.h (hunk 3 above)
Comment (tlauda) [reply src/platform/baytrail/include/platform/lib/clk.h]: @lgirdwood I don't think we should forward it globally. Every header should declare things for itself.

File: src/include/sof/drivers/edma.h [hunk 4]
```diff
@@ -73,8 +73,6 @@
 #define EDMA_TCD_ATTR_DSIZE_32BYTE	0x0005
```
Comment (dbaluta) [inline src/include/sof/drivers/edma.h]: @tlauda what dmac stands for? dma controller?

Comment (tlauda) [reply src/include/sof/drivers/edma.h]: @dbaluta I would assume so.

**Comment by tlauda** [comment]
@lgirdwood Comment regards sof structure alignment added.

File: src/include/sof/sof.h (hunk 1 above)
Comment (tlauda) [reply src/include/sof/sof.h]: @lgirdwood Done.

File: src/include/sof/lib/dma.h [hunk 5]
```diff
@@ -216,6 +216,11 @@ struct dma_chan_data {
 	void *private;
 };
```
Comment (dbaluta) [inline src/include/sof/lib/dma.h]: @tlauda would appreciate a short explanation of *why* we are doing this change. E.g Move it to header because it is needed by multiple modules, etc.

File: src/include/sof/drivers/edma.h (hunk 4 above)
Comment (dbaluta) [reply src/include/sof/drivers/edma.h]: Ok, I'm fine with this change.

File: src/platform/baytrail/include/platform/lib/clk.h (hunk 3 above)
Comment (lgirdwood) [reply src/platform/baytrail/include/platform/lib/clk.h]: I'm thinking along the lines of linux/kernel.h, but this can be looked at later.

File: src/include/sof/lib/dma.h (hunk 5 above)
Comment (tlauda) [reply src/include/sof/lib/dma.h]: @dbaluta Done. Also updated similar dai commit.

File: src/lib/pm_runtime.c [hunk 6]
```diff
@@ -21,8 +21,6 @@ static struct pm_runtime_data *prd;
 
 void pm_runtime_init(void)
 {
-	trace_pm("pm_runtime_init()");
-
```
Comment (paulstelian97) [inline src/lib/pm_runtime.c]: Maybe verbose trace? Or in case of debugging future PM-related issues such prints would be added by whoever is doing the debugging?

File: src/lib/agent.c [hunk 7]
```diff
@@ -67,25 +65,24 @@ void sa_init(struct sof *sof, uint64_t timeout)
 
 	trace_sa("sa_init(), timeout = %u", timeout);
 
-	sa = rzalloc(SOF_MEM_ZONE_SYS, 0, SOF_MEM_CAPS_RAM, sizeof(*sa));
-	sof->sa = sa;
+	sof->sa = rzalloc(SOF_MEM_ZONE_SYS, 0, SOF_MEM_CAPS_RAM,
```
Comment (paulstelian97) [inline src/lib/agent.c]: I assume this structure either was already there or was added by some other patch?

File: src/lib/pm_runtime.c (hunk 6 above)
Comment (tlauda) [reply src/lib/pm_runtime.c]: @paulstelian97 Initialization is somewhat logged by trace_point, so there is no need for additional log.

File: src/include/sof/lib/alloc.h [hunk 8]
```diff
@@ -255,4 +255,10 @@ void free_heap(enum mem_zone zone);
 void heap_trace_all(int force);
 void heap_trace(struct mm_heap *heap, int size);
 
+/* retrieve memory map pointer */
+static inline struct mm *memmap_get(void)
+{
+	return sof_get()->memory_map;
```
Comment (paulstelian97) [inline src/include/sof/lib/alloc.h]: I feel this commit could be squashed with the previous? I'm not strongly in favor of keeping vs squashing so it's your choice.

File: src/lib/agent.c (hunk 7 above)
Comment (tlauda) [reply src/lib/agent.c]: Yes, it was here before.

File: src/include/sof/lib/alloc.h (hunk 8 above)
Comment (tlauda) [reply src/include/sof/lib/alloc.h]: @paulstelian97 I am fan of limiting the number of commits, but there were many requests in the past to keep some changes in separate commits, so I'm trying to match those requests.

File: src/lib/pm_runtime.c (hunk 6 above)
Comment (paulstelian97) [reply src/lib/pm_runtime.c]: Guess I have to implement trace_point at some point (maybe reuse the same memory location for panics as well, as I don't have those implemented on i.MX either). Fair enough then.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by tlauda** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
Looks like CI scan is down.
